### PR TITLE
[Product] Product variant validation

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | no |
| Related tickets |  |
| License         | MIT |

### CODE REVIEW

File: features/product/managing_product_variants/product_variant_validation.feature [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        But I do not set its price
+        And I try to add it
+        Then I should be notified that price is required
+        And the "VODKA_WYBOROWA_PREMIUM" variant of the "Wyborowa Vodka" product should not appear in the shop
+
+    @ui
+    Scenario: Adding a new product variant with price lower then 0
+        Given I want to create a new variant of this product
+        When I specify its code as "VODKA_WYBOROWA_PREMIUM"
+        And I set its price to "$-80.00"
```
Comment (pamil) [inline features/product/managing_product_variants/product_variant_validation.feature]: `$-80.00` -> `-$80.00`

File: src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php [hunk 2]
```diff
@@ -58,6 +78,26 @@ public function choosePricingCalculator($name)
 
     /**
      * {@inheritdoc}
+     *
+     * @throws ElementNotFoundException
+     */
+    public function getValidationMessageForForm()
+    {
+        $foundElement = $this->getDocument()->find('xpath', '//form[@name = \'sylius_product_variant\']');
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php]: `find('xpath', '//form[@name = \'sylius_product_variant\']')` -> `find('css', 'form[name="sylius_product_variant"]')`

Comment (pamil) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePage.php]: `$foundElement` -> `$formElement`

File: src/Sylius/Behat/Page/Admin/ProductVariant/CreatePageInterface.php [hunk 3]
```diff
@@ -24,6 +24,13 @@
      * @param int $price
      */
     public function specifyPrice($price);
+    /**
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/ProductVariant/CreatePageInterface.php]: Missing blank line above.

File: src/Sylius/Bundle/ProductBundle/spec/Validator/UniqueSimpleProductCodeValidatorSpec.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+use Sylius\Component\Product\Model\ProductInterface;
+use Sylius\Component\Product\Model\ProductVariantInterface;
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Symfony\Component\Validator\Violation\ConstraintViolationBuilderInterface;
+
+final class UniqueSimpleProductCodeValidatorSpec extends ObjectBehavior
+{
+    function let(ExecutionContextInterface $context,
```
Comment (pamil) [inline src/Sylius/Bundle/ProductBundle/spec/Validator/UniqueSimpleProductCodeValidatorSpec.php]: Weird formatting going on here.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 5]
```diff
@@ -206,6 +223,35 @@ public function theProductVariantShouldAppearInTheShop($productVariantCode, Prod
     }
 
     /**
+     * @Then the :productVariantCode variant of the :product product should not appear in the shop
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: shop -> store

Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: I would prefer to keep consistency with previous [step](https://github.com/Sylius/Sylius/pull/6768/files#diff-9caa6c41a42784d0658dcbb0dbc447eaR213)

Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: We should use `store` in both cases to be consistent with all the other steps 😄

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 6]
```diff
@@ -206,6 +223,35 @@ public function theProductVariantShouldAppearInTheShop($productVariantCode, Prod
...
+    /**
+     * @Then the :product product should have no variants
+     */
+    public function theProductShouldHaveNoVariants(ProductInterface $product)
+    {
+        $this->assertNumberOfVariantsOnProductPage($product, 0);
+    }
+
+    /**
+     * @Then the :product product should have only one 1 variant
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: only **one 1** variant - we should use either `1` or `one` 😄

File: src/Sylius/Bundle/ProductBundle/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -195,6 +195,11 @@ private function addResourcesSection(ArrayNodeDefinition $node)
                                             ->defaultValue(['sylius'])
                                             ->cannotBeEmpty()
                                         ->end()
+                                        ->arrayNode('generation')
+                                            ->prototype('scalar')->end()
+                                            ->defaultValue(['sylius'])
+                                            ->cannotBeEmpty()
+                                        ->end()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/DependencyInjection/Configuration.php]: Do we need this?

File: src/Sylius/Bundle/ProductBundle/Validator/UniqueSimpleProductCodeValidator.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+{
+    /**
+     * @var ProductVariantRepositoryInterface
+     */
+    private $productVariantRepository;
+
+    /**
+     * @param ProductVariantRepositoryInterface $productVariantRepository
+     */
+    public function __construct(ProductVariantRepositoryInterface $productVariantRepository) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/Validator/UniqueSimpleProductCodeValidator.php]: Bracket in a wrong place 😆

**Comment by michalmarcinkowski** [comment]
Good job Łukasz! Please apply my comments in a separate PR.

**Pull request merged**